File: app/src/main/java/com/example/milen/hope/CreateList.java
package com.example.milen.hope;

import android.content.Intent;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.LinearLayout;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.util.LinkedList;
import java.util.List;

public class CreateList extends AppCompatActivity {

    private static final String TAG = CreateList.class.getName();
    private static final int RC_SIGN_IN = 4444;
    GoogleSignInClient mGoogleSignInClient;
    GoogleSignInAccount account;
    private DatabaseReference mDatabase;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_create_list);


        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestEmail()
                .build();
        mGoogleSignInClient = GoogleSignIn.getClient(this, gso);

        FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                plusButttonClick();
            }
        });

        account = GoogleSignIn.getLastSignedInAccount(CreateList.this);
        if (account == null)
            signInUser();
    }

    public void plusButttonClick() {
        LinearLayout layout = (LinearLayout) findViewById(R.id.items_layout);
        LinearLayout.LayoutParams lparams = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.FILL_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT);
        EditText editText = new EditText(getApplicationContext());
        editText.setLayoutParams(lparams);
        editText.setHint("To do Item");
        layout.addView(editText);
    }

    public void signInUser() {
        Intent signInIntent = mGoogleSignInClient.getSignInIntent();
        startActivityForResult(signInIntent, RC_SIGN_IN);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        // Result returned from launching the Intent from GoogleSignInClient.getSignInIntent(...);
        if (requestCode == RC_SIGN_IN) {
            // The Task returned from this call is always completed, no need to attach
            // a listener.
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);

            try {
                account = task.getResult(ApiException.class);

                // Signed in successfully, show authenticated UI.
            } catch (ApiException e) {
                // The ApiException status code indicates the detailed failure reason.
                // Please refer to the GoogleSignInStatusCodes class reference for more information.
                Log.w(TAG, "signInResult:failed code=" + e.getStatusCode());
            }

        }
    }

    public void onCreateClick(View view) {
        EditText listNameBox = findViewById(R.id.list_name_box);
        String listName = listNameBox.getText().toString();

        List<ToDoItem> toDoItems = new LinkedList<>();
        LinearLayout layout = findViewById(R.id.items_layout);
        for (int index = 0; index < layout.getChildCount(); index++) {
            View textView = layout.getChildAt(index);
            if ((textView instanceof EditText) && (textView.getId() != R.id.list_name_box)) {
                String itemText = ((EditText) textView).getText().toString();
                toDoItems.add(new ToDoItem(itemText));
            }
        }

        String userID = account.getId();

        mDatabase = FirebaseDatabase.getInstance().getReference();
        mDatabase.child("users").child(userID).child(listName).push();

        for (ToDoItem item: toDoItems) {
            mDatabase.child("users").child(userID).child(listName).child(item.getDescription())
                    .setValue(item.isCompleted());
        }
        finish();
    }
}

File: app/src/main/java/com/example/milen/hope/CreateList.kt
package com.example.milen.hope

import android.content.Intent
import android.support.design.widget.FloatingActionButton
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.EditText
import android.widget.LinearLayout

import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.tasks.Task
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase

import java.util.LinkedList

class CreateList : AppCompatActivity() {
    internal var mGoogleSignInClient: GoogleSignInClient? = null
    internal var account: GoogleSignInAccount? = null
    private var mDatabase: DatabaseReference? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_create_list)


        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestEmail()
                .build()
        mGoogleSignInClient = GoogleSignIn.getClient(this, gso)

        val fab = findViewById<FloatingActionButton>(R.id.fab)
        fab.setOnClickListener { plusButttonClick() }

        account = GoogleSignIn.getLastSignedInAccount(this@CreateList)
        if (account == null)
            signInUser()
    }

    fun plusButttonClick() {
        val layout = findViewById<View>(R.id.items_layout) as LinearLayout
        val lparams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.FILL_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)
        val editText = EditText(applicationContext)
        editText.layoutParams = lparams
        editText.hint = "To do Item"
        layout.addView(editText)
    }

    fun signInUser() {
        val signInIntent = mGoogleSignInClient!!.signInIntent
        startActivityForResult(signInIntent, RC_SIGN_IN)
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        // Result returned from launching the Intent from GoogleSignInClient.getSignInIntent(...);
        if (requestCode == RC_SIGN_IN) {
            // The Task returned from this call is always completed, no need to attach
            // a listener.
            val task = GoogleSignIn.getSignedInAccountFromIntent(data)

            try {
                account = task.getResult<ApiException>(ApiException::class.java)

                // Signed in successfully, show authenticated UI.
            } catch (e: ApiException) {
                // The ApiException status code indicates the detailed failure reason.
                // Please refer to the GoogleSignInStatusCodes class reference for more information.
                Log.w(TAG, "signInResult:failed code=" + e.statusCode)
            }

        }
    }

    fun onCreateClick(view: View) {
        val listNameBox = findViewById<EditText>(R.id.list_name_box)
        val listName = listNameBox.text.toString()

        val toDoItems = LinkedList<ToDoItem>()
        val layout = findViewById<LinearLayout>(R.id.items_layout)
        for (index in 0 until layout.childCount) {
            val textView = layout.getChildAt(index)
            if (textView is EditText && textView.getId() != R.id.list_name_box) {
                val itemText = textView.text.toString()
                toDoItems.add(ToDoItem(itemText))
            }
        }

        val userID = account!!.id

        mDatabase = FirebaseDatabase.getInstance().reference
        mDatabase!!.child("users").child(userID!!).child(listName).push()

        for (item in toDoItems) {
            mDatabase!!.child("users").child(userID).child(listName).child(item.description)
                    .setValue(item.isCompleted)
        }
        finish()
    }

    companion object {

        private val TAG = CreateList::class.java.name
        private val RC_SIGN_IN = 4444
    }
}
